interfaces/builtin/shared_memory.go

```diff
@@ -0,0 +1,229 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
+
+const sharedMemoryBaseDeclarationPlugs = `
+  shared-memory:
+    allow-installation: true
+    allow-connection:
+      slot-attributes:
+        shared-memory: $PLUG(shared-memory)
+    allow-auto-connection:
+      slot-publisher-id:
+        - $PLUG_PUBLISHER_ID
+      slot-attributes:
+        shared-memory: $PLUG(shared-memory)
+`
+
+const sharedMemoryBaseDeclarationSlots = `
+  shared-memory:
+    allow-installation: false
+    deny-connection: true
+    deny-auto-connection: true
+`
+
+func validateSharedMemoryPath(path string) error {
+	if len(path) == 0 {
+		return fmt.Errorf("shared-memory interface path is empty")
+	}
+
+	if strings.TrimSpace(path) != path {
```

Comment:
should we also do a check using `cleanSubPath` right? I suppose it's unnecessary for now since we disallow `/` entirely, but maybe when we support the long-form we should add that check too

Reply:
Actually I realized that I was still allowing ".." as path; now I added `cleanSubPath`, and that fixed it. Thanks for pointing that out :-)